pkg/cluster/parse.go

```diff
@@ -0,0 +1,210 @@
+package cluster
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/yaml"
+
+	anywherev1 "github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+)
+
+type parsed struct {
+	objects        objects
+	cluster        *anywherev1.Cluster
+	datacenter     apiObject
+	machineConfigs []runtime.Object
+}
+
+type basicAPIObject struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+}
+
+func (k *basicAPIObject) empty() bool {
+	return k.APIVersion == "" && k.Kind == ""
+}
+
+type apiObject interface {
+	runtime.Object
+	GetName() string
+}
+
+func keyForObject(o apiObject) string {
+	return key(o.GetObjectKind().GroupVersionKind().GroupVersion().String(), o.GetObjectKind().GroupVersionKind().Kind, o.GetName())
+}
+
+type objects map[string]apiObject
+
+func (o objects) add(obj apiObject) {
+	o[keyForObject(obj)] = obj
+}
+
+func (o objects) getFromRef(apiVersion string, ref anywherev1.Ref) apiObject {
+	return o[keyForRef(apiVersion, ref)]
+}
+
+func key(apiVersion, kind, name string) string {
+	// this assumes we don't allow to have objects in multiple namespaces
+	return fmt.Sprintf("%s%s%s", apiVersion, kind, name)
+}
+
+func keyForRef(apiVersion string, ref anywherev1.Ref) string {
+	return key(apiVersion, ref.Kind, ref.Name)
+}
+
+func ParseConfigFromFile(path string) (*Config, error) {
+	content, err := os.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("reading cluster config file: %v", err)
+	}
+
+	return ParseConfig(content)
+}
+
+func ParseConfig(yamlManifest []byte) (*Config, error) {
+	parsed := &parsed{
+		objects: objects{},
+	}
+	yamlObjs := strings.Split(string(yamlManifest), "---")
+
+	for _, yamlObj := range yamlObjs {
+		k := &basicAPIObject{}
+		err := yaml.Unmarshal([]byte(yamlObj), k)
+		if err != nil {
+			return nil, err
+		}
+
+		// Ignore empty objects.
+		// Empty objects are generated if there are weird things in manifest files like e.g. two --- in a row without a yaml doc in the middle
+		if k.empty() {
+			continue
+		}
+
+		var obj apiObject
+
+		switch k.Kind {
+		case anywherev1.ClusterKind:
+			if parsed.cluster != nil {
+				return nil, errors.New("only one Cluster per yaml manifest is allowed")
+			}
+			parsed.cluster = &anywherev1.Cluster{}
+			obj = parsed.cluster
+		case anywherev1.VSphereDatacenterKind:
```

Comment:
This is gonna grow so it would be good if we could make it dynamic and extensible by retrieving a marshalling func from a registry and asking the func to perform the unmarshall and return the `apiObj` so it can be cached.